src/ring_theory/polynomial_algebra.lean

```diff
@@ -0,0 +1,342 @@
+variables [ring A] [algebra R A]
+
+-- The instance set up in `data.polynomial` allows `[comm_semiring R]` and `[semiring A]`.
+-- Here we provide a specialisation, as otherwise some typeclass inference problems below
+-- cause deterministic timeouts. Suggestions for better fixes welcome.
+instance algebra_of_algebra' : algebra R (polynomial A) := polynomial.algebra_of_algebra
+
+namespace polynomial_equiv_tensor
+
+/--
+(Implementation detail).
+The bare function underlying `A ⊗[R] polynomial R →ₐ[R] polynomial A`, on pure tensors.
+-/
+def to_fun (a : A) (p : polynomial R) : polynomial A :=
+p.sum (λ n r, monomial n (a * algebra_map R A r))
+
+/--
+(Implementation detail).
+The function underlying `A ⊗[R] polynomial R →ₐ[R] polynomial A`,
+as a linear map in the second factor.
+-/
+def to_fun_linear_right (a : A) : polynomial R →ₗ[R] polynomial A :=
+{ to_fun := to_fun R A a,
+  map_smul' := λ r p,
+  begin
+    dsimp [to_fun],
+    rw finsupp.sum_smul_index,
+    { dsimp [finsupp.sum],
+      rw finset.smul_sum,
+      apply finset.sum_congr rfl,
+      intros k hk,
+      rw [monomial_eq_smul_X, monomial_eq_smul_X, algebra.smul_def, ← C_mul', ← C_mul',
+          ← mul_assoc],
+      congr' 1,
+      rw [← algebra.commutes, ← algebra.commutes],
+      simp only [ring_hom.map_mul, polynomial.algebra_map_apply, mul_assoc] },
+    { intro i, simp only [ring_hom.map_zero, mul_zero, monomial_zero_right] },
+  end,
+  map_add' := λ p q,
+  begin
+    simp only [to_fun],
+    rw finsupp.sum_add_index,
+    { simp only [monomial_zero_right, forall_const, ring_hom.map_zero, mul_zero], },
+    { intros i r s, simp only [ring_hom.map_add, mul_add, monomial_add], },
+  end, }
+
+/--
+(Implementation detail).
+The function underlying `A ⊗[R] polynomial R →ₐ[R] polynomial A`,
+as a bilinear function of two arguments.
+-/
+def to_fun_bilinear : A →ₗ[R] polynomial R →ₗ[R] polynomial A :=
+{ to_fun := to_fun_linear_right R A,
+  map_smul' := by {
+    intros, unfold to_fun_linear_right,
+    congr, simp only [linear_map.coe_mk],
+    unfold to_fun finsupp.sum monomial,
+    simp_rw [finset.smul_sum, finsupp.smul_single,  ← algebra.smul_mul_assoc],
+    refl },
+  map_add' := by {
+    intros, unfold to_fun_linear_right,
+    congr, simp only [linear_map.coe_mk],
+    unfold to_fun finsupp.sum monomial,
+    simp_rw [← finset.sum_add_distrib, ← finsupp.single_add, ← add_mul],
+    refl } }
+
+/--
+(Implementation detail).
+The function underlying `A ⊗[R] polynomial R →ₐ[R] polynomial A`,
+as a linear map.
+-/
+def to_fun_linear : A ⊗[R] polynomial R →ₗ[R] polynomial A :=
+tensor_product.lift (to_fun_bilinear R A)
+
+-- We apparently need to provide the decidable instance here
+-- in order to successfully rewrite by this lemma.
+lemma to_fun_linear_mul_tmul_mul_aux_1
+  (p : polynomial R) (k : ℕ) (h : decidable (¬p.coeff k = 0)) (a : A) :
+  ite (¬coeff p k = 0) (a * (algebra_map R A) (coeff p k)) 0 = a * (algebra_map R A) (coeff p k) :=
+begin
+  haveI := h,
```

Comment:
```suggestion
```

Anything that's left of the colon is already available as an instance, no matter whether it's in brackets or parens.